File: everyDay/1090.受标签影响的最大值.cpp
/*
 * @lc app=leetcode.cn id=1090 lang=cpp
 *
 * [1090] 受标签影响的最大值
 */

// @lc code=start
class Solution {
public:
    int vis[100000];
    int largestValsFromLabels(vector<int>& values, vector<int>& labels, int numWanted, int useLimit) {
        // 排好队一个一个
        vector<pair<int, int> > V;
        for (int i = 0; i < values.size(); i++) {
            V.push_back({values[i], labels[i]});
            // V.emplace_back({values[i], labels[i]});
        }
        int sum = 0;
        sort(V.begin(), V.end(), greater<pair<int, int> >());
        for (int i = 0; i < V.size(); i++) {
            // cout << V[i].first << endl;
            if (!numWanted) {
                break;
            }
            if (vis[V[i].second] < useLimit) {
                vis[V[i].second]++;
                sum += V[i].first;
                numWanted--;
            }
        }

        return sum;
    }
};
// @lc code=end

File: everyDay/1090.受标签影响的最大值.go
/*
 * @lc app=leetcode.cn id=1090 lang=golang
 *
 * [1090] 受标签影响的最大值
 */

// @lc code=start
func largestValsFromLabels(values []int, labels []int, numWanted int, useLimit int) int {
	// make a new arr to sort
	// sort by idx
	n := len(values)
	idx := make([]int, n)

	m := make(map[int]int)
	for i := 0; i < n; i++ {
		idx[i] = i
	}
	sort.Slice(idx, func(i, j int) bool {
		return values[idx[i]] > values[idx[j]]
	})
	sum := 0
	for i := 0; i < n; i++ {
		if numWanted == 0 {
			break
		}
		l := labels[idx[i]]
		v := values[idx[i]]
		if m[l] < useLimit {
			numWanted--;
			m[l]++
			sum += v
		}
	}
	return sum

}
// @lc code=end

